# Esame
#### Daniele Nucciarelli

In [1]:
import pandas as pd

Lettura del csv dei dati

In [2]:
import pandas as pd
df = pd.read_csv("/home/jovyan/work/data/dataset_exam.csv", index_col=0)

In [3]:
df.head()

City        lat        lng  Population  Density  Surface   
0  Acquasparta  42.691173  12.546451        4611     57.0    81.61  \
1     Allerona  42.812033  11.973053        1722     21.0    82.61   
2      Alviano  42.588208  12.295936        1428     60.0    23.90   
3       Amelia  42.553528  12.416776       11852     89.0   132.50   
4       Arrone  42.583433  12.769865        2703     66.0    41.04   

   Depriv_idx  CovidCases_jan_jun_2020  MaxHospitalized_jan_jun_2020   
0    1.013507                        6                             1  \
1   -1.309272                        3                             1   
2   -2.148882                        2                             1   
3   -0.832084                        5                             1   
4   -0.479919                        7                             1   

   AvgHospitalized_jan_jun_2020  ...  mean_pm10_ug/m3_std_2019   
0                      0.273585  ...                  7.288500  \
1                      0.336735  ...                  7.038712   
2                      0.252632  ...                  8.908002   
3                      0.317308  ...                  7.749981   
4                      0.475728  ...                  7.416702   

   mean_pm10_ug/m3_median_2019  max_pm10_ug/m3_median_2019   
0                    14.674647                   27.591324  \
1                    14.752206                   26.312595   
2                    19.626043                   30.008255   
3                    16.706606                   31.103338   
4                    12.388708                   26.191957   

   max_pm10_ug/m3_mean_2019  max_pm10_ug/m3_std_2019   
0                 31.958118                15.960667  \
1                 30.565736                18.320679   
2                 34.685780                17.674601   
3                 36.154165                19.364122   
4                 32.194737                20.364874   

   min_pm10_ug/m3_median_2019  min_pm10_ug/m3_mean_2019   
0                    6.754477                  7.535171  \
1                    4.790663                  5.623008   
2                    8.948166                  9.722944   
3                    7.095291                  7.938031   
4                    3.970692                  4.888365   

   min_pm10_ug/m3_std_2019                 Zone  Region  
0                 4.595696  collinare e montana  Umbria  
1                 3.754214  collinare e montana  Umbria  
2                 5.046609  collinare e montana  Umbria  
3                 4.636137  collinare e montana  Umbria  
4                 4.001205  collinare e montana  Umbria  

[5 rows x 36 columns]

In [4]:
df.columns

Index(['City', 'lat', 'lng', 'Population', 'Density', 'Surface', 'Depriv_idx',
       'CovidCases_jan_jun_2020', 'MaxHospitalized_jan_jun_2020',
       'AvgHospitalized_jan_jun_2020', 'Deceased_jan_jun_2020',
       'MaxIntensiveCare_jan_jun_2020', 'AvgIntensiveCare_jan_jun_2020',
       'mean_pm10_ug/m3_mean_jan_jun_2020', 'mean_pm10_ug/m3_std_jan_jun_2020',
       'mean_pm10_ug/m3_median_jan_jun_2020',
       'max_pm10_ug/m3_mean_jan_jun_2020', 'max_pm10_ug/m3_std_jan_jun_2020',
       'max_pm10_ug/m3_median_jan_jun_2020',
       'min_pm10_ug/m3_mean_jan_jun_2020', 'min_pm10_ug/m3_std_jan_jun_2020',
       'min_pm10_ug/m3_median_jan_jun_2020',
       'mean_pm10_ug/m3_mean_jan_jun_2020.1',
       'mean_pm10_ug/m3_std_jan_jun_2020.1',
       'mean_pm10_ug/m3_median_jan_jun_2020.1', 'mean_pm10_ug/m3_mean_2019',
       'mean_pm10_ug/m3_std_2019', 'mean_pm10_ug/m3_median_2019',
       'max_pm10_ug/m3_median_2019', 'max_pm10_ug/m3_mean_2019',
       'max_pm10_ug/m3_std_2019', 'min_pm10_ug/

In [5]:
df.describe()

lat        lng     Population     Density     Surface   
count  92.000000  92.000000      92.000000   92.000000   92.000000  \
mean   42.905182  12.472221    9568.315217   89.459565   92.003804   
std     0.273627   0.292700   22012.983364  108.091455   96.297586   
min    42.403124  11.973053      96.000000    2.390000   10.000000   
25%    42.703714  12.234268    1573.750000   32.000000   35.735000   
50%    42.873417  12.481221    2946.500000   66.500000   58.935000   
75%    43.073766  12.718593    6913.500000  108.000000   97.267500   
max    43.548980  13.092928  166969.000000  787.000000  525.780000   

       Depriv_idx  CovidCases_jan_jun_2020  MaxHospitalized_jan_jun_2020   
count   92.000000                92.000000                     92.000000  \
mean    -0.455099                15.054348                      2.673913   
std      1.357775                39.688447                      6.293392   
min     -5.251338                 0.000000                      0.000000   
25%     -1.204914                 0.000000                      0.000000   
50%     -0.492604                 5.000000                      1.000000   
75%      0.351280                11.000000                      3.000000   
max      2.872051               332.000000                     51.000000   

       AvgHospitalized_jan_jun_2020  Deceased_jan_jun_2020  ...   
count                     92.000000              92.000000  ...  \
mean                       0.801545               0.826087  ...   
std                        1.988844               2.094518  ...   
min                        0.000000               0.000000  ...   
25%                        0.000000               0.000000  ...   
50%                        0.185896               0.000000  ...   
75%                        0.853774               1.000000  ...   
max                       14.971698              14.000000  ...   

       mean_pm10_ug/m3_median_jan_jun_2020.1  mean_pm10_ug/m3_mean_2019   
count                              92.000000                  92.000000  \
mean                               14.475831                  16.430487   
std                                 3.014723                   2.620349   
min                                 7.420084                  10.744741   
25%                                12.673209                  14.608065   
50%                                15.098714                  16.623700   
75%                                16.377223                  18.087417   
max                                21.254544                  22.373268   

       mean_pm10_ug/m3_std_2019  mean_pm10_ug/m3_median_2019   
count                 92.000000                    92.000000  \
mean                   7.788327                    14.892218   
std                    0.571131                     2.769346   
min                    6.939437                     8.817261   
25%                    7.387661                    12.719118   
50%                    7.648703                    15.328800   
75%                    8.057968                    16.689253   
max                    9.674192                    20.719354   

       max_pm10_ug/m3_median_2019  max_pm10_ug/m3_mean_2019   
count                   92.000000                 92.000000  \
mean                    26.885255                 31.688390   
std                      4.958767                  5.724870   
min                     16.463018                 20.667668   
25%                     24.180037                 28.689235   
50%                     26.361343                 30.668131   
75%                     28.995770                 34.065552   
max                     48.968746                 56.453095   

       max_pm10_ug/m3_std_2019  min_pm10_ug/m3_median_2019   
count                92.000000                   92.000000  \
mean                 17.816801                    6.811227   
std                   4.342591                    2.446697   
min                  11.

## Pre-processing




### Valori mancanti

In [6]:
pd.isnull(df).sum()

City                                     0
lat                                      0
lng                                      0
Population                               0
Density                                  0
Surface                                  0
Depriv_idx                               0
CovidCases_jan_jun_2020                  0
MaxHospitalized_jan_jun_2020             0
AvgHospitalized_jan_jun_2020             0
Deceased_jan_jun_2020                    0
MaxIntensiveCare_jan_jun_2020            0
AvgIntensiveCare_jan_jun_2020            0
mean_pm10_ug/m3_mean_jan_jun_2020        0
mean_pm10_ug/m3_std_jan_jun_2020         0
mean_pm10_ug/m3_median_jan_jun_2020      0
max_pm10_ug/m3_mean_jan_jun_2020         0
max_pm10_ug/m3_std_jan_jun_2020          0
max_pm10_ug/m3_median_jan_jun_2020       0
min_pm10_ug/m3_mean_jan_jun_2020         0
min_pm10_ug/m3_std_jan_jun_2020          0
min_pm10_ug/m3_median_jan_jun_2020       0
mean_pm10_ug/m3_mean_jan_jun_2020.1      0
mean_pm10_u

Sono presenti dei valori mancanti nelle colonne Region e min_pm10_ug/m3_std_2019.

#### Valore mancante nella colonna Region


In [7]:
df.loc[df["Region"].isnull()]

City        lat        lng  Population  Density  Surface  Depriv_idx   
80  Spello  42.990945  12.671775        8464    137.0    61.65   -0.730392  \

    CovidCases_jan_jun_2020  MaxHospitalized_jan_jun_2020   
80                        7                             2  \

    AvgHospitalized_jan_jun_2020  ...  mean_pm10_ug/m3_std_2019   
80                      0.608247  ...                  7.861205  \

    mean_pm10_ug/m3_median_2019  max_pm10_ug/m3_median_2019   
80                    15.591546                   27.833431  \

    max_pm10_ug/m3_mean_2019  max_pm10_ug/m3_std_2019   
80                 32.547284                18.056519  \

    min_pm10_ug/m3_median_2019  min_pm10_ug/m3_mean_2019   
80                    4.850987                  6.058144  \

    min_pm10_ug/m3_std_2019      Zone  Region  
80                  4.58133  di valle     NaN  

[1 rows x 36 columns]

Spello è un comune umbro; in effetti anche tutti gli altri campionamenti sono stati effettuati in Umbria:

In [8]:
df.groupby('Region').Region.count()

Region
Umbria    91
Name: Region, dtype: int64

Pertanto si può eliminare questa colonna perché non aggiunge informazione allo specifico campionamento

In [9]:
df.drop('Region', axis=1, inplace=True)

#### Valore mancante nella colonna min_pm10_ug/m3_std_2019

I valori di questa colonna rappresentano il valore minimo espresso in μg/m3 della deviazione standard giornaliera di pm10 rilevato nel 2019

In [10]:
df.loc[df["min_pm10_ug/m3_std_2019"].isnull()]

City        lat        lng  Population  Density   
45  Monte_Castello_di_Vibio  42.840694  12.352286        1504     47.0  \

    Surface  Depriv_idx  CovidCases_jan_jun_2020   
45    31.95   -1.138733                        0  \

    MaxHospitalized_jan_jun_2020  AvgHospitalized_jan_jun_2020  ...   
45                             0                           0.0  ...  \

    mean_pm10_ug/m3_mean_2019  mean_pm10_ug/m3_std_2019   
45                  17.468114                  7.535464  \

    mean_pm10_ug/m3_median_2019  max_pm10_ug/m3_median_2019   
45                    16.200014                   26.999504  \

    max_pm10_ug/m3_mean_2019  max_pm10_ug/m3_std_2019   
45                 30.651951                15.562025  \

    min_pm10_ug/m3_median_2019  min_pm10_ug/m3_mean_2019   
45                    8.564888                  9.239437  \

    min_pm10_ug/m3_std_2019                 Zone  
45                      NaN  collinare e montana  

[1 rows x 35 columns]

Droppare l'intera entry mi sembra esagerato. Fillarla con un valore... potrei metterci il valore del 2020, oppure una media delle std degli altri comuni, oppure lasciare NaN

#### Eliminare colonne inutili

Si ricordi il numero di entries del dataset:

In [11]:
len(df)

92

Sono presenti delle colonne \(nominate "*.1"\) le cui descrizioni non sono presenti nella feature description. Con il sospetto che si tratti di colonne duplicate, si confrontano queste colonne con altre e si nota che:

In [12]:
(df["mean_pm10_ug/m3_mean_jan_jun_2020"]==df["mean_pm10_ug/m3_mean_jan_jun_2020.1"]).sum()

92

In [13]:
(df["mean_pm10_ug/m3_std_jan_jun_2020"]==df["mean_pm10_ug/m3_std_jan_jun_2020.1"]).sum()

92

In [14]:
(df["mean_pm10_ug/m3_median_jan_jun_2020"]==df["mean_pm10_ug/m3_median_jan_jun_2020.1"]).sum()

92

ovvero per ognuna delle tre colonne, la pd.Series data dal confronto tra due colonne potenzialmente coincidenti è costituita da 92 True, pari al numero di entries.
Pertanto si possono rimuovere le suddette colonne:

In [16]:
df.drop(
    columns=[
    "mean_pm10_ug/m3_mean_jan_jun_2020.1",
    "mean_pm10_ug/m3_std_jan_jun_2020.1",
    "mean_pm10_ug/m3_median_jan_jun_2020.1"
    ], 
    inplace=True
)

In conclusione, il pd.Dataframe pre-processato presenta 4 colonne in meno del pd.Dataframe di partenza:

In [20]:
df.shape

(92, 32)

### Plots

3. Creare due grafici che mettono in relazioni variabili relative a demografia, epidemiologia e geografia del territorio.

  3.1 (**Extra**)

  * inserire nelle tooltip il nome della città;
  * creare un plot dove la terza feature da mettere in relazione è selezionabile a piacere tra le colonne del data frame. Sugli assi principali si trovano rispettivamente Population e Density.

4. Correlazione: studia la correlazione lineare tra le feature

5. Maps (**Extra**): creare due esempi di grafici su mappa con delle colonne e punti che rappresentano delle informazioni sul territorio.

  5.1 *Show Density*: la colonna rappresenta la densità. Questa viene confermata dal colore, rosso dove la densità è più alta. Il colore utilizza le colonne Population e Surface.

  Le colonne devono prensentare come informazioni aggiuntive i seguenti dettagli:
  * Città
  * Densità
  * Popolazione 
  * Superfice

  5.2 *Show CovidCases*: la colonna rappresenta il numero di casi Covid. Il colore viene composto in base ai decessi, il numero medio di casi intensivi e numero medio di casi ospitalizzati. I decessi danno alla colonna un colore rosso.

  Le colonne devono prensentare come informazioni aggiuntive i seguenti dettagli:
  * Città
  * \# casi covid
  * avg. intensive care 
  * avg. hospitalized




## Modeling

### Unsupervised Learning

#### Clustering

6. Utilizzare KMeans per creare dei cluster partendo da un numero pari a 2 fino ad un massimo di 18. Utilizzare le seguenti feature:

  * Depriv_idx
  * CovidCases_jan_jun_2020 / Population
  * MaxHospitalized_jan_jun_2020 / Population
  * AvgHospitalized_jan_jun_2020 / Population
  * Deceased_jan_jun_2020 / Population
  * MaxIntensiveCare_jan_jun_2020 / Population
  * AvgIntensiveCare_jan_jun_2020 / Population
  * max_pm10_ug/m3_mean_jan_jun_2020
  * max_pm10_ug/m3_std_jan_jun_2020
  * max_pm10_ug/m3_median_jan_jun_2020
  * mean_pm10_ug/m3_mean_jan_jun_2020
  * mean_pm10_ug/m3_std_jan_jun_2020
  * mean_pm10_ug/m3_median_jan_jun_2020
  * min_pm10_ug/m3_mean_jan_jun_2020
  * min_pm10_ug/m3_std_jan_jun_2020
  * min_pm10_ug/m3_median_jan_jun_2020

  6.2 Fare un grafico dell'inerzia all'aumentare del numero di cluster.

  6.3 Maps (**Extra**): proiettare i risultati su mappa  utilizzando `lat` e `lng` e aggiungendo la feature `Zone` come categoria.



#### Clustering con PCA (**extra**)

7. Effettua lo stesso studio, ma utilizzando le componenti principali anzichè le feature pure, in modo tale da avere rappresentazioni bi- o tridimensionali contenenti più informazione possibile


### Supervised Learning

#### Regression

8. Utilizzare due metodi per effettuare una regressione della feature `CovidCases_jan_jun_2020 / Population`
avendo come input le seguenti colonne:

  * lat
  * lng
  * Surface
  * mean_pm10_ug/m3_mean_jan_jun_2020
  * mean_pm10_ug/m3_std_jan_jun_2020
  * mean_pm10_ug/m3_median_jan_jun_2020

  8.1 Verificare quindi l'addestramento.

  8.2 **Extra**: 
  * fare un grafico del test set durante la fase di training
  * fare un grafico della predizione del test set